File: algo/src/main/java/com/hartwig/actin/algo/evaluation/priortumor/PreviousTumorRuleMapper.java
package com.hartwig.actin.algo.evaluation.priortumor;

import java.time.LocalDate;
import java.util.Map;

import com.google.common.collect.Maps;
import com.hartwig.actin.algo.evaluation.FunctionCreator;
import com.hartwig.actin.algo.evaluation.RuleMapper;
import com.hartwig.actin.algo.evaluation.RuleMappingResources;
import com.hartwig.actin.treatment.datamodel.EligibilityRule;

import org.jetbrains.annotations.NotNull;

public class PreviousTumorRuleMapper extends RuleMapper {

    public PreviousTumorRuleMapper(@NotNull final RuleMappingResources resources) {
        super(resources);
    }

    @NotNull
    @Override
    public Map<EligibilityRule, FunctionCreator> createMappings() {
        Map<EligibilityRule, FunctionCreator> map = Maps.newHashMap();

        map.put(EligibilityRule.HAS_ACTIVE_SECOND_MALIGNANCY, hasActiveSecondMalignancyCreator());
        map.put(EligibilityRule.HAS_HISTORY_OF_SECOND_MALIGNANCY, hasHistoryOfSecondMalignancyCreator());
        map.put(EligibilityRule.HAS_HISTORY_OF_SECOND_MALIGNANCY_BELONGING_TO_DOID_X, hasHistoryOfSecondMalignancyWithDoidCreator());
        map.put(EligibilityRule.HAS_HISTORY_OF_SECOND_MALIGNANCY_BELONGING_TO_DOID_TERM_X,
                hasHistoryOfSecondMalignancyWithDoidTermCreator());
        map.put(EligibilityRule.HAS_HISTORY_OF_SECOND_MALIGNANCY_WITHIN_X_YEARS, hasHistoryOfSecondMalignancyWithinYearsCreator());

        return map;
    }

    @NotNull
    private FunctionCreator hasActiveSecondMalignancyCreator() {
        return function -> new HasActiveSecondMalignancy();
    }

    @NotNull
    private FunctionCreator hasHistoryOfSecondMalignancyCreator() {
        return function -> new HasHistoryOfSecondMalignancy();
    }

    @NotNull
    private FunctionCreator hasHistoryOfSecondMalignancyWithDoidCreator() {
        return function -> {
            String doidToMatch = functionInputResolver().createOneStringInput(function);
            return new HasHistoryOfSecondMalignancyWithDoid(doidModel(), doidToMatch);
        };
    }

    @NotNull
    private FunctionCreator hasHistoryOfSecondMalignancyWithDoidTermCreator() {
        return function -> {
            // TODO Map DOID term to DOID
            String doidTermToMatch = functionInputResolver().createOneDoidTermInput(function);
            return new HasHistoryOfSecondMalignancyWithDoid(doidModel(), doidTermToMatch);
        };
    }

    @NotNull
    private FunctionCreator hasHistoryOfSecondMalignancyWithinYearsCreator() {
        return function -> {
            int maxYears = functionInputResolver().createOneIntegerInput(function);
            LocalDate minDate = referenceDateProvider().date().minusYears(maxYears);

            return new HasHistoryOfSecondMalignancyWithinYears(minDate);
        };
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/priortumor/PreviousTumorRuleMapper.kt
package com.hartwig.actin.algo.evaluation.priortumor

import com.hartwig.actin.algo.evaluation.FunctionCreator
import com.hartwig.actin.algo.evaluation.RuleMapper
import com.hartwig.actin.algo.evaluation.RuleMappingResources
import com.hartwig.actin.datamodel.trial.DoidTermParameter
import com.hartwig.actin.datamodel.trial.EligibilityFunction
import com.hartwig.actin.datamodel.trial.IntegerParameter
import com.hartwig.actin.datamodel.trial.ManyDoidTermsParameter
import com.hartwig.actin.datamodel.trial.Parameter
import com.hartwig.actin.trial.input.EligibilityRule

class PreviousTumorRuleMapper(resources: RuleMappingResources) : RuleMapper(resources) {

    override fun createMappings(): Map<EligibilityRule, FunctionCreator> {
        return mapOf(
            EligibilityRule.HAS_ACTIVE_SECOND_MALIGNANCY to hasActiveSecondMalignancyCreator(),
            EligibilityRule.HAS_HISTORY_OF_SECOND_MALIGNANCY to hasHistoryOfSecondMalignancyCreator(),
            EligibilityRule.HAS_HISTORY_OF_SECOND_MALIGNANCY_IGNORING_DOID_TERMS_X to hasHistoryOfSecondMalignancyIgnoringSomeDoidsCreator(),
            EligibilityRule.HAS_HISTORY_OF_SECOND_MALIGNANCY_BELONGING_TO_DOID_TERM_X to hasHistoryOfSecondMalignancyWithDoidTermCreator(),
            EligibilityRule.HAS_HISTORY_OF_SECOND_MALIGNANCY_WITHIN_X_YEARS to hasHistoryOfSecondMalignancyWithinYearsCreator(),
            EligibilityRule.HAS_HISTORY_OF_SECOND_MALIGNANCY_WITHIN_X_YEARS_IGNORING_DOID_TERMS_Y
                    to hasHistoryOfSecondMalignancyWithinYearsIgnoringSomeDoidTermsCreator(),
        )
    }

    private fun hasActiveSecondMalignancyCreator(): FunctionCreator {
        return { HasActiveSecondMalignancy(evaluationLabels.priorTumor) }
    }

    private fun hasHistoryOfSecondMalignancyCreator(): FunctionCreator {
        return { HasHistoryOfSecondMalignancy(evaluationLabels.priorTumor) }
    }

    private fun hasHistoryOfSecondMalignancyIgnoringSomeDoidsCreator(): FunctionCreator {
        return { function: EligibilityFunction ->
            val doidInputsToIgnore = function.param<ManyDoidTermsParameter>(0).value
            val doidsToIgnore = doidInputsToIgnore.map { doidModel.toDoid(it) }.toSet()
            HasHistoryOfSecondMalignancyIgnoringDoidTerms(doidModel, doidsToIgnore, null, evaluationLabels.priorTumor)
        }
    }

    private fun hasHistoryOfSecondMalignancyWithDoidTermCreator(): FunctionCreator {
        return { function: EligibilityFunction ->
            val doidInputToMatch = function.param<DoidTermParameter>(0).value
            val doidToMatch = doidModel.toDoid(doidInputToMatch)
            HasHistoryOfSecondMalignancyWithDoid(doidModel, doidToMatch, evaluationLabels.priorTumor)
        }
    }

    private fun hasHistoryOfSecondMalignancyWithinYearsCreator(): FunctionCreator {
        return { function: EligibilityFunction ->
            val maxYears = function.param<IntegerParameter>(0).value
            val minDate = referenceDateProvider.date().minusYears(maxYears.toLong())
            HasHistoryOfSecondMalignancyWithinYears(minDate, evaluationLabels.priorTumor)
        }
    }

    private fun hasHistoryOfSecondMalignancyWithinYearsIgnoringSomeDoidTermsCreator(): FunctionCreator {
        return { function: EligibilityFunction ->
            function.expectTypes(Parameter.Type.INTEGER, Parameter.Type.MANY_DOID_TERMS)
            val maxYears = function.param<IntegerParameter>(0).value
            val doidInputsToIgnore = function.param<ManyDoidTermsParameter>(1).value
            val doidsToIgnore = doidInputsToIgnore.map { doidModel.toDoid(it) }.toSet()
            val minDate = referenceDateProvider.date().minusYears(maxYears.toLong())
            HasHistoryOfSecondMalignancyIgnoringDoidTerms(doidModel, doidsToIgnore, minDate, evaluationLabels.priorTumor)
        }
    }
}
